extmod/network_wiznet5k_lwip.c

```diff
@@ -0,0 +1,570 @@
+    #else
+    mp_handle_pending(true);
+    #endif
+}
+
+STATIC mp_obj_t mpy_wiznet_read_int(mp_obj_t none_in) {
+    (void)none_in;
+    wizchip_clrinterrupt(IK_SOCK_0);
+    #if _WIZCHIP_ == W5100S
+    setSn_IR(0, Sn_IR_RECV);
+    #endif
+
+    // Handle incoming data
+    wiznet5k_try_poll();
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(mpy_wiznet_read_int_obj, mpy_wiznet_read_int);
+
+STATIC void wiznet5k_config_interrupt(bool enabled) {
+    if (!wiznet5k_obj.use_interrupt) {
+        return;
+    }
+    mp_hal_pin_interrupt(
+        wiznet5k_obj.pin_intn,
+        MP_OBJ_FROM_PTR(&mpy_wiznet_read_int_obj),
+        (enabled)? MP_HAL_PIN_TRIGGER_FALL : MP_HAL_PIN_TRIGGER_NONE,
+        true
+        );
+}
+
+STATIC void wiz_spi_read(uint8_t *buf, uint16_t len) {
+    uint8_t *tx_buf = alloca(len);
+    ((mp_machine_spi_p_t *)machine_spi_type.protocol)->transfer((mp_obj_base_t *)wiznet5k_obj.spi,len, tx_buf,buf);
+}
+
+STATIC void wiz_spi_write(const uint8_t *buf, uint16_t len) {
+    ((mp_machine_spi_p_t *)machine_spi_type.protocol)->transfer((mp_obj_base_t *)wiznet5k_obj.spi,len,buf, NULL);
+}
+
+STATIC uint8_t wiz_spi_readbyte() {
+    uint8_t tx_buf;
+    uint8_t rx_buf;
+    ((mp_machine_spi_p_t *)machine_spi_type.protocol)->transfer((mp_obj_base_t *)wiznet5k_obj.spi,1, &tx_buf, &rx_buf);
+    return rx_buf;
+}
+
+STATIC void wiz_spi_writebyte(const uint8_t buf) {
+    ((mp_machine_spi_p_t *)machine_spi_type.protocol)->transfer((mp_obj_base_t *)wiznet5k_obj.spi,1, &buf, NULL);
+}
+
+STATIC void wiznet5k_init(void) {
+    mp_hal_pin_output(wiznet5k_obj.cs);
+    mp_hal_pin_output(wiznet5k_obj.rst);
+
+    // Reset the chip
+    mp_hal_pin_low(wiznet5k_obj.rst);
+    mp_hal_delay_ms(1); // datasheet says 2us
+    mp_hal_pin_high(wiznet5k_obj.rst);
+    mp_hal_delay_ms(150); // datasheet says 150ms
+
+    // Set physical interface callbacks
+    reg_wizchip_cris_cbfunc(wiz_cris_enter, wiz_cris_exit);
+    reg_wizchip_cs_cbfunc(wiz_cs_select, wiz_cs_deselect);
+    reg_wizchip_spi_cbfunc(wiz_spi_readbyte, wiz_spi_writebyte);
+    reg_wizchip_spiburst_cbfunc(wiz_spi_read, wiz_spi_write);
+
+    // Configure 16k buffers for fast MACRAW
+    uint8_t sn_size[16] = {16, 0, 0, 0, 0, 0, 0, 0, 16, 0, 0, 0, 0, 0, 0, 0};
+    ctlwizchip(CW_INIT_WIZCHIP, sn_size);
+
+    // Seems we need a small delay after init
+    mp_hal_delay_ms(250);
+
+    // If the device doesn't have a MAC address then set one
+    uint8_t mac[6];
+    getSHAR(mac);
+    if ((mac[0] | mac[1] | mac[2] | mac[3] | mac[4] | mac[5]) == 0) {
+        mp_hal_get_mac(MP_HAL_MAC_ETH0, mac);
+        setSHAR(mac);
+    }
+
+    if (wiznet5k_obj.use_interrupt) {
+        // Enable Wiznet interrupts for socket 0
+        wizchip_setinterruptmask(IK_SOCK_0);
+        #if _WIZCHIP_ == W5100S
+        // Enable data receive interrupt
+        setSn_IMR(0, Sn_IR_RECV);
+        // Enable interrupt pin
+        setMR(MR2_G_IEN);
+        #endif
```

Comment:
for W5500 the `setSn_IMR(...)` is needed, but not the `setMR(...)`; I'd suggest making the first command unconditional (ie not guarded) and the second one guarded by W5100S